test/remote/redis_tests.js

```diff
@@ -0,0 +1,240 @@
+    })
+  })
+
+  describe('concurrent sets:', () => {
+    before(() => {
+      return P.all(connections.map((connection, index) => connection.set('foo', `${index}`)))
+    })
+
+    it('data was set', () => {
+      return connections[0].get('foo')
+        .then(result => assert.ok(result === '0' || result === '1'))
+    })
+  })
+
+  describe('connection.update:', () => {
+    before(() => {
+      return connections[0].update('foo', oldValue => `${oldValue}2`)
+    })
+
+    it('first connection reads data', () => {
+      return connections[0].get('foo')
+        .then(result => assert.equal(result[1], '2'))
+    })
+
+    it('second connection reads data', () => {
+      return connections[1].get('foo')
+        .then(result => assert.equal(result[1], '2'))
+    })
+  })
+
+  describe('update non-existent key:', () => {
+    before(() => {
+      return connections[0].update('wibble', () => 'blee')
+    })
+
+    it('data was updated', () => {
+      return connections[1].get('wibble')
+        .then(result => assert.equal(result, 'blee'))
+    })
+  })
+
+  describe('update existing key to falsey value:', () => {
+    before(() => {
+      return connections[0].update('wibble', () => '')
+    })
+
+    it('data was updated', () => {
+      return connections[1].get('wibble')
+        .then(result => assert.ok(result === null))
+    })
+  })
+
+  describe('concurrent updates of the same key:', () => {
+    const errors = []
+    let successIndex
+
+    before(() => {
+      return P.all(
+        connections.map(
+          (connection, index) => {
+            return connection.update('foo', () => `${index}`)
+              .then(() => successIndex = index)
+              .catch(error => errors.push(error))
+          }
+        )
+      )
+    })
+
+    it('one update failed', () => {
+      assert.equal(errors.length, 1)
+      assert.equal(errors[0].message, 'Unspecified error')
+      assert.equal(errors[0].errno, 999)
+    })
+
+    it('the other update completed successfully', () => {
+      return connections[0].get('foo')
+        .then(result => assert.equal(result, successIndex))
+    })
+  })
+
+  describe('concurrent updates of different keys:', () => {
+    before(() => {
+      return P.all([
+        connections[0].update('foo', () => 'bar'),
+        connections[1].update('baz', () => 'qux')
+      ])
+    })
+
+    it('first update completed successfully', () => {
+      return connections[1].get('foo')
+        .then(result => assert.equal(result, 'bar'))
+    })
+
+    it('second update completed successfully', () => {
+      return connections[0].get('baz')
+        .then(result => assert.equal(result, 'qux'))
+    })
+  })
+
+  describe('reentrant updates of different keys:', () => {
+    let error
+
+    before(() => {
+      return P.all([
+        connections[0].update('foo', oldValue => `${oldValue}2`),
+        connections[0].update('baz', oldValue => `${oldValue}2`).catch(e => error = e)
```

Comment:
Ditto here re: nondeterminism producing sporadic test failures.  Could we guarantee the necessary ordering of events by doing the second `.update()` inside the callback of the first?  Or, are we already guaranteed a correct ordering by virtue of the way the node event loop works?